crates/hotshot-stake-table/src/vec_based.rs

```diff
@@ -0,0 +1,495 @@
+    /// The stake table used for leader election.
+    last_epoch_start: StakeTableSnapshot<K1, K2>,
+
+    /// Total stakes for different versions
+    head_total_stake: U256,
+    epoch_start_total_stake: U256,
+    last_epoch_start_total_stake: U256,
+
+    /// We only support committing the finalized versions.
+    /// Commitment for a finalized version is a triple where
+    ///  - First item is the rescue hash of the bls keys
+    ///  - Second item is the rescue hash of the Schnorr keys
+    ///  - Third item is the rescue hash of all the stake amounts
+    epoch_start_comm: (F, F, F),
+    last_epoch_start_comm: (F, F, F),
+
+    /// The mapping from public keys to their location in the Merkle tree.
+    #[serde(skip)]
+    bls_mapping: HashMap<K1, usize>,
+}
+
+impl<K1, K2, F> StakeTableScheme for StakeTable<K1, K2, F>
+where
+    K1: Eq + Hash + Clone + ToFields<F>,
+    K2: Eq + Hash + Clone + ToFields<F>,
+    F: RescueParameter,
+{
+    /// The stake table is indexed by BLS key
+    type Key = K1;
+    /// The auxiliary information is the associated Schnorr key
+    type Aux = K2;
+    type Amount = U256;
+    type Commitment = (F, F, F);
+    type LookupProof = ();
+    // TODO(Chengyu): Can we make it references?
+    type IntoIter = <Vec<(K1, U256, K2)> as ark_std::iter::IntoIterator>::IntoIter;
+
+    fn register(
+        &mut self,
+        new_key: Self::Key,
+        amount: Self::Amount,
+        aux: Self::Aux,
+    ) -> Result<(), StakeTableError> {
+        match self.bls_mapping.get(&new_key) {
+            Some(_) => Err(StakeTableError::ExistingKey),
+            None => {
+                let pos = self.bls_mapping.len();
+                self.head.bls_keys.push(new_key.clone());
+                self.head.schnorr_keys.push(aux);
+                self.head.stake_amount.push(amount);
+                self.head_total_stake += amount;
+                self.bls_mapping.insert(new_key, pos);
+                Ok(())
+            }
+        }
+    }
+
+    fn deregister(&mut self, existing_key: &Self::Key) -> Result<(), StakeTableError> {
+        match self.bls_mapping.get(existing_key) {
+            Some(pos) => {
+                self.head_total_stake -= self.head.stake_amount[*pos];
+                self.head.stake_amount[*pos] = U256::zero();
+                Ok(())
+            }
+            None => Err(StakeTableError::KeyNotFound),
+        }
+    }
+
+    fn commitment(&self, version: SnapshotVersion) -> Result<Self::Commitment, StakeTableError> {
+        match version {
+            // IMPORTANT: we don't support committing the head version b/c it's not finalized.
+            SnapshotVersion::Head => Err(StakeTableError::SnapshotUnsupported),
+            SnapshotVersion::EpochStart => Ok(self.epoch_start_comm),
+            SnapshotVersion::LastEpochStart => Ok(self.last_epoch_start_comm),
+            SnapshotVersion::BlockNum(_) => Err(StakeTableError::SnapshotUnsupported),
```

Comment:
the first and last cases can be simplified as: `_ => Err()` (all other cases)

Reply:
changed in acfdc82